src/python/pants/pantsd/service/scheduler_service.py

```diff
@@ -65,34 +65,29 @@ def _process_event_queue(self):
     except Queue.Empty:
       return
 
-    with self.lock:
-      try:
-        subscription, is_initial_event, files = (event['subscription'],
-                                                 event['is_fresh_instance'],
-                                                 [f.decode('utf-8') for f in event['files']])
-      except (KeyError, UnicodeDecodeError) as e:
-        self._logger.warn('%r raised by invalid watchman event: %s', e, event)
-        return
+    try:
+      subscription, is_initial_event, files = (event['subscription'],
+                                               event['is_fresh_instance'],
+                                               [f.decode('utf-8') for f in event['files']])
+    except (KeyError, UnicodeDecodeError) as e:
+      self._logger.warn('%r raised by invalid watchman event: %s', e, event)
+      return
 
-      self._logger.debug('processing {} files for subscription {} (first_event={})'
-                         .format(len(files), subscription, is_initial_event))
+    self._logger.debug('processing {} files for subscription {} (first_event={})'
+                       .format(len(files), subscription, is_initial_event))
 
-      # The first watchman event is a listing of all files - ignore it.
-      if not is_initial_event:
-        self._handle_batch_event(files)
+    # The first watchman event is a listing of all files - ignore it.
+    if not is_initial_event:
+      self._handle_batch_event(files)
 
-    if not self._ready.is_set(): self._ready.set()
     self._event_queue.task_done()
 
   def warm_product_graph(self, spec_roots):
     """Runs an execution request against the captive scheduler given a set of input specs to warm.
 
     :returns: A `LegacyGraphHelper` instance for graph construction.
     """
-    # Block warming until the initial watchman filesystem event is seen.
-    self._ready.wait()
```

Comment:
As discussed in #engine, I think your idea to block if the graph is non-empty (ie, `length > 0`) would be a good middle ground.